### ISSUE TITLE
Error when launching instance inside of VPC 

### ISSUE THREAD

**Issue by gwilton** [infochimps-labs/ironfan#232]
Hello All,

I upgraded to 4.7.1 for the purpose of launching instances inside of VPC. It seems to be broken. You can find the details below.

I just finished testing the same configuration on 4.3.0 and the instance is launched inside of the VPC with this version of Ironfan but it gets stuck when trying to SSH. I am assuming this is due to needing an Elastic IP.

Can anyone provide some feedback on the best way of integrating Ironfan with AWS VPC. What would be the best version to use for this and proper configuration. I been using Ironfan for over a year now. VPC is now a requirement for me due to security reasons. Thank you.

**Ironfan Output:**

bundle exec knife cluster launch sandbox
Inventorying servers in sandbox cluster, all facets, all servers
  sandbox:          Loading chef
  sandbox:          Loading ec2
  sandbox:          Reconciling DSL and provider information
  +----------------+-------+-------------+----------+------------+-----+
  | Name           | Chef? | State       | Flavor   | AZ         | Env |
  +----------------+-------+-------------+----------+------------+-----+
  | sandbox-raid-0 | no    | not running | m1.large | us-east-1c | dev |
  +----------------+-------+-------------+----------+------------+-----+
Syncing to chef
Preparing shared resources:
  sandbox:          Loading chef
  sandbox:          Loading ec2
  sandbox:          Reconciling DSL and provider information
Loaded information for 1 computer(s) in cluster sandbox
  sandbox:          ensuring security group permissions
  sandbox:            ensuring access from sandbox to sandbox
  vpc-fe8fca94:raid-vpc:      ensuring udp access from 0.0.0.0/0 to 500..500
  vpc-fe8fca94:raid-vpc:      ensuring udp access from 0.0.0.0/0 to 4500..4500
  vpc-fe8fca94:ssh:   ensuring tcp access from 0.0.0.0/0 to 22..22

Launching computers
  +----------------+-------+-------------+----------+------------+-----+
  | Name           | Chef? | State       | Flavor   | AZ         | Env |
  +----------------+-------+-------------+----------+------------+-----+
  | sandbox-raid-0 | no    | not running | m1.large | us-east-1c | dev |
  +----------------+-------+-------------+----------+------------+-----+
  sandbox-raid-0:   creating cloud machine
WARNING: Error running #<Ironfan::Broker::Computer(server=#<Ironfan::Dsl::Server(name="0", run_list_items=c{ role[systemwide], role[chef_client], role[ssh], role[nfs_client], role[set_hostname], role[volumes], role[package_set], role[minidash], role[org_base], role[org_users], role[org_final], role[tuning], volumes::build_raid, role[sandbox_cluster], role[sandbox_raid] }, clouds=c{ ec2 }, volumes=c{ ephemeral0, ephemeral1, md0 }, environment=:dev, cluster_role=#<Ironfan::Dsl::Role>, facet_role=#Ironfan::Dsl::Role, cluster_name="sandbox", facet_name="raid")>, resources=c{ keypair, security_group__vpc-fe8fca94:systemwide, security_group__vpc-fe8fca94:ssh, security_group__vpc-fe8fca94:raid-vpc }, drives=c{ ephemeral0, ephemeral1, md0, root }, providers=c{ chef, iaas })>:
WARNING: undefined method `group_id' for nil:NilClass
ERROR: undefined method`group_id' for nil:NilClass (NoMethodError)
/usr/local/rvm/gems/ruby-1.9.3-p194@homebase/gems/ironfan-4.7.1/lib/ironfan/provider/ec2/machine.rb:244:in `block in launch_description'
/usr/local/rvm/gems/ruby-1.9.3-p194@homebase/gems/ironfan-4.7.1/lib/ironfan/provider/ec2/machine.rb:242:in`map'
/usr/local/rvm/gems/ruby-1.9.3-p194@homebase/gems/ironfan-4.7.1/lib/ironfan/provider/ec2/machine.rb:242:in `launch_description'
/usr/local/rvm/gems/ruby-1.9.3-p194@homebase/gems/ironfan-4.7.1/lib/ironfan/provider/ec2/machine.rb:155:in`create!'
/usr/local/rvm/gems/ruby-1.9.3-p194@homebase/gems/ironfan-4.7.1/lib/ironfan/broker/computer.rb:77:in `launch'
/usr/local/rvm/gems/ruby-1.9.3-p194@homebase/gems/ironfan-4.7.1/lib/ironfan/broker/computer.rb:324:in`block in launch'
/usr/local/rvm/gems/ruby-1.9.3-p194@homebase/gems/ironfan-4.7.1/lib/ironfan.rb:35:in `block (3 levels) in parallel'
/usr/local/rvm/gems/ruby-1.9.3-p194@homebase/gems/ironfan-4.7.1/lib/ironfan.rb:134:in`safely'
/usr/local/rvm/gems/ruby-1.9.3-p194@homebase/gems/ironfan-4.7.1/lib/ironfan.rb:34:in `block (2 levels) in parallel'
ERROR: /usr/local/rvm/gems/ruby-1.9.3-p194@homebase/gems/ironfan-4.7.1/lib/ironfan/provider/ec2/machine.rb:244:in`block in launch_description'
/usr/local/rvm/gems/ruby-1.9.3-p194@homebase/gems/ironfan-4.7.1/lib/ironfan/provider/ec2/machine.rb:242:in `map'
/usr/local/rvm/gems/ruby-1.9.3-p194@homebase/gems/ironfan-4.7.1/lib/ironfan/provider/ec2/machine.rb:242:in`launch_description'
/usr/local/rvm/gems/ruby-1.9.3-p194@homebase/gems/ironfan-4.7.1/lib/ironfan/provider/ec2/machine.rb:155:in `create!'
/usr/local/rvm/gems/ruby-1.9.3-p194@homebase/gems/ironfan-4.7.1/lib/ironfan/broker/computer.rb:77:in`launch'
/usr/local/rvm/gems/ruby-1.9.3-p194@homebase/gems/ironfan-4.7.1/lib/ironfan/broker/computer.rb:324:in `block in launch'
/usr/local/rvm/gems/ruby-1.9.3-p194@homebase/gems/ironfan-4.7.1/lib/ironfan.rb:35:in`block (3 levels) in parallel'
/usr/local/rvm/gems/ruby-1.9.3-p194@homebase/gems/ironfan-4.7.1/lib/ironfan.rb:134:in `safely'
/usr/local/rvm/gems/ruby-1.9.3-p194@homebase/gems/ironfan-4.7.1/lib/ironfan.rb:34:in`block (2 levels) in parallel'
WARNING: Error launching #<NoMethodError: undefined method `group_id' for nil:NilClass>; skipping after-launch tasks.
Applying aggregations:
  sandbox:          Loading chef
  sandbox:          Loading ec2
  sandbox:          Reconciling DSL and provider information
Loaded information for 1 computer(s) in cluster sandbox
  +----------------+-------+-------------+----------+------------+-----+
  | Name           | Chef? | State       | Flavor   | AZ         | Env |
  +----------------+-------+-------------+----------+------------+-----+
  | sandbox-raid-0 | yes   | not running | m1.large | us-east-1c | dev |
  +----------------+-------+-------------+----------+------------+-----+

**Cluster Config:**
# 
# Sandbox cluster -- use this for general development
# 

Ironfan.cluster 'sandbox' do
  cloud(:ec2) do
    permanent           false
    availability_zones ['us-east-1c']
    flavor              'm1.large'
    backing             'ebs'
    image_name          'ironfan-natty'
    bootstrap_distro    'ubuntu10.04-ironfan'
    chef_client_script  'client.rb'
    mount_ephemerals
  end

  environment           :dev

  role                  :systemwide
  cloud(:ec2).security_group :systemwide
  role                  :chef_client
  role                  :ssh
  cloud(:ec2).security_group(:ssh).authorize_port_range 22..22
  role                  :nfs_client
  #cloud(:ec2).security_group :nfs_client
  role                  :set_hostname

  role                  :volumes
  role                  :package_set,   :last
  role                  :minidash,      :last

  role                  :org_base
  role                  :org_users
  role                  :org_final,     :last

  role                  :tuning,        :last

  facet :raid do
    instances           1
    cloud(:ec2).flavor        'm1.large'
    recipe              'volumes::build_raid', :first

```
cloud(:ec2) do
  availability_zones ['us-east-1c']
  vpc                 'vpc-fe8fca94'       # set this to your existing vpc instance id
  subnet              'subnet-a08ecbca'    # set this to your existing subnet id
  security_group      'raid-vpc' do  
    authorize_port_range 500,"0.0.0.0/0",'udp'
    authorize_port_range 4500,"0.0.0.0/0",'udp'
  end
end

cloud(:ec2).mount_ephemerals
raid_group(:md0) do
  device            '/dev/md0'
  mount_point       '/raid0'
  level             0
  sub_volumes       [:ephemeral0, :ephemeral1]
end
```

  end

  cluster_role.override_attributes({
    })
end

**Comment by UncleTallest** [comment]
If the problem truly is/was related to the lack of an Elastic IP; you'll need to add an `elastic_ip '<elastic.ip.address.goes.here>'` statement to your existing `cloud(:ec2) do` block within your facet definition. You need specify a pre-allocated Elastic IP.

About the second; I don't know where it is but between these two lines:

``` ruby
recipe 'volumes::build_raid', :first

cloud(:ec2) do
```

You seem to have an extraneous back-tick (unshifted tilde on a US keyboard) or single quote between those two lines causing ruby to add a trailing single quote to the method name group_id (about line 244 in machine.rb); hence the  'unknown method' `group_id` error. I can't see this as I am inferring it from the change in formatting in the two instances you've given that I can see.

**Comment by gwilton** [comment]
@schade Thanks for the reply. 

The group_id error was not due to the single quote, i just didn't use the github markdown properly in my original post (First post). However, after testing from a clean cloned homebase the group_id error is now gone. 

I added the elastic_ip, but still dealing with some issues due to allocation_id. This is what I am seeing. Is it possible that the versions between chef, fog, ironfan are not playing nice with each other. Can you let me know with what versions you managed to get VPC fully working? A full working example VPC cluster configuration can help me as well. Thanks!

At the moment all I can due via Ironfan is launch an instances into an existing VPC. I am failing at associating EIP and running ssh, cluster commands on the facet. Your help is really appreciated.

[homebase  (master)]$ gem list | grep ironfan
**ironfan (4.7.1)**

[homebase  (master)]$ _knife cluster list_
  +---------+----------------------------------------------------------------+
  | cluster | path                                                           |
  +---------+----------------------------------------------------------------+
  | sandbox | ~/Documents/workspace/homebase/clusters/sandbox.rb |
  +---------+----------------------------------------------------------------+

[homebase  (master)]$ _knife cluster show sandbox_
Inventorying servers in sandbox cluster, all facets, all servers
  sandbox:          Loading chef
  sandbox:          Loading ec2
  sandbox:          Reconciling DSL and provider information
  +----------------+-------+-------------+----------+------------+-----+
  | Name           | Chef? | State       | Flavor   | AZ         | Env |
  +----------------+-------+-------------+----------+------------+-----+
  | sandbox-raid-0 | no    | not running | m1.large | us-east-1c | dev |
  +----------------+-------+-------------+----------+------------+-----+

**I can launch into a VPC but elastic IP is not being associated due to allocation id**

[homebase  (master)]$ _knife cluster launch sandbox_
Inventorying servers in sandbox cluster, all facets, all servers
  sandbox:          Loading chef
  sandbox:          Loading ec2
  sandbox:          Reconciling DSL and provider information
  +----------------+-------+-------------+----------+------------+-----+
  | Name           | Chef? | State       | Flavor   | AZ         | Env |
  +----------------+-------+-------------+----------+------------+-----+
  | sandbox-raid-0 | no    | not running | m1.large | us-east-1c | dev |
  +----------------+-------+-------------+----------+------------+-----+
Syncing to chef
Preparing shared resources:
  sandbox:          Loading chef
  sandbox:          Loading ec2
  sandbox:          Reconciling DSL and provider information
Loaded information for 1 computer(s) in cluster sandbox
  sandbox:          ensuring security group permissions
  sandbox:            ensuring access from sandbox to sandbox
  vpc-fe8fca94:raid-vpc:      ensuring tcp access from 0.0.0.0/0 to 22..22
  vpc-fe8fca94:ssh:   ensuring tcp access from 0.0.0.0/0 to 22..22

Launching computers
  +----------------+-------+-------------+----------+------------+-----+
  | Name           | Chef? | State       | Flavor   | AZ         | Env |
  +----------------+-------+-------------+----------+------------+-----+
  | sandbox-raid-0 | no    | not running | m1.large | us-east-1c | dev |
  +----------------+-------+-------------+----------+------------+-----+
  sandbox-raid-0:   creating cloud machine

  i-7732be06:       tagging with {"cluster"=>"sandbox", "facet"=>"raid", "index"=>0, "name"=>"sandbox-raid-0", "Name"=>"sandbox-raid-0"}
  vol-cb15b9b5:     tagging with {"cluster"=>"sandbox", "facet"=>"raid", "index"=>0, "name"=>"sandbox-raid-0-root", "Name"=>"sandbox-raid-0-root", "server"=>"sandbox-raid-0", "mount_point"=>"/", "device"=>"/dev/sda1"}
  sandbox-raid-0:   setting termination flag false
  sandbox-raid-0:   associating Elastic IP 107.23.163.84

``` ruby
WARNING: Error running :
WARNING: InvalidParameterCombination => You must specify an allocation id when mapping an address to a VPC instance
ERROR: InvalidParameterCombination => You must specify an allocation id when mapping an address to a VPC instance (Fog::Compute::AWS::Error)
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/excon-0.16.10/lib/excon/connection.rb:292:in `request_kernel'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/excon-0.16.10/lib/excon/connection.rb:103:in `request'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/fog-1.8.0/lib/fog/core/connection.rb:21:in `request'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/fog-1.8.0/lib/fog/aws/compute.rb:383:in `_request'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/fog-1.8.0/lib/fog/aws/compute.rb:378:in `request'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/fog-1.8.0/lib/fog/aws/requests/compute/associate_address.rb:35:in `associate_address'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/ironfan-4.7.1/lib/ironfan/provider/ec2/elastic_ip.rb:58:in `block (2 levels) in save!'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/ironfan-4.7.1/lib/ironfan.rb:134:in `safely'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/ironfan-4.7.1/lib/ironfan/provider/ec2/elastic_ip.rb:57:in `block in save!'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/ironfan-4.7.1/lib/ironfan.rb:172:in `unless_dry_run'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/ironfan-4.7.1/lib/ironfan/provider/ec2/elastic_ip.rb:56:in `save!'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/ironfan-4.7.1/lib/ironfan/broker/computer.rb:96:in `block in save'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/ironfan-4.7.1/lib/ironfan/broker/computer.rb:96:in `each'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/ironfan-4.7.1/lib/ironfan/broker/computer.rb:96:in `save'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/ironfan-4.7.1/lib/ironfan/broker/computer.rb:79:in `launch'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/ironfan-4.7.1/lib/ironfan/broker/computer.rb:324:in `block in launch'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/ironfan-4.7.1/lib/ironfan.rb:35:in `block (3 levels) in parallel'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/ironfan-4.7.1/lib/ironfan.rb:134:in `safely'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/ironfan-4.7.1/lib/ironfan.rb:34:in `block (2 levels) in parallel'
ERROR: /usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/excon-0.16.10/lib/excon/connection.rb:292:in `request_kernel'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/excon-0.16.10/lib/excon/connection.rb:103:in `request'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/fog-1.8.0/lib/fog/core/connection.rb:21:in `request'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/fog-1.8.0/lib/fog/aws/compute.rb:383:in `_request'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/fog-1.8.0/lib/fog/aws/compute.rb:378:in `request'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/fog-1.8.0/lib/fog/aws/requests/compute/associate_address.rb:35:in `associate_address'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/ironfan-4.7.1/lib/ironfan/provider/ec2/elastic_ip.rb:58:in `block (2 levels) in save!'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/ironfan-4.7.1/lib/ironfan.rb:134:in `safely'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/ironfan-4.7.1/lib/ironfan/provider/ec2/elastic_ip.rb:57:in `block in save!'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/ironfan-4.7.1/lib/ironfan.rb:172:in `unless_dry_run'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/ironfan-4.7.1/lib/ironfan/provider/ec2/elastic_ip.rb:56:in `save!'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/ironfan-4.7.1/lib/ironfan/broker/computer.rb:96:in `block in save'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/ironfan-4.7.1/lib/ironfan/broker/computer.rb:96:in `each'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/ironfan-4.7.1/lib/ironfan/broker/computer.rb:96:in `save'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/ironfan-4.7.1/lib/ironfan/broker/computer.rb:79:in `launch'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/ironfan-4.7.1/lib/ironfan/broker/computer.rb:324:in `block in launch'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/ironfan-4.7.1/lib/ironfan.rb:35:in `block (3 levels) in parallel'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/ironfan-4.7.1/lib/ironfan.rb:134:in `safely'
/usr/local/rvm/gems/ruby-1.9.3-p194@ihb/gems/ironfan-4.7.1/lib/ironfan.rb:34:in `block (2 levels) in parallel'
```

  sandbox-raid-0:   syncing EBS volumes
  sandbox-raid-0:   launching
  sandbox-raid-0:   waiting for ready
  sandbox-raid-0:   trying ssh

**Since EIP was not assign, I assigned it manually**

[homebase  (master)]$ telnet 107.23.163.84 22
Trying 107.23.163.84..

Connected to ec2-107-23-163-84.compute-1.amazonaws.com.
Escape character is '^]'.
SSH-2.0-OpenSSH_5.3p1 Debian-3ubuntu7
Protocol mismatch.
Connection closed by foreign host.

[homebase  (master)]$ knife cluster show sandbox
Inventorying servers in sandbox cluster, all facets, all servers
  sandbox:          Loading chef
  sandbox:          Loading ec2
  sandbox:          Reconciling DSL and provider information
  +----------------+-------+---------+----------+------------+-----+------------+---------------+-------------+------------+
  | Name           | Chef? | State   | Flavor   | AZ         | Env | MachineID  | Public IP     | Private IP  | Created On |
  +----------------+-------+---------+----------+------------+-----+------------+---------------+-------------+------------+
  | sandbox-raid-0 | yes   | running | m1.large | us-east-1c | dev | i-7732be06 | 107.23.163.84 | 10.14.3.194 | 2013-01-08 |
  +----------------+-------+---------+----------+------------+-----+------------+---------------+-------------+------------+

**I then try to ssh and then i get a FATAL error which seems to be coming from knife ssh**

[homebase  (master)]$ knife cluster ssh sandbox uptime
Inventorying servers in sandbox cluster, all facets, all servers
  sandbox:          Loading chef
  sandbox:          Loading ec2
  sandbox:          Reconciling DSL and provider information
**FATAL: No nodes returned from search!**

**Below is the cluster config I am using for this**

``` ruby
#
# Sandbox cluster -- use this for general development
#
Ironfan.cluster 'sandbox' do
  cloud(:ec2) do
    permanent           false
    availability_zones ['us-east-1c']
    flavor              'm1.large'
    backing             'ebs'
    image_name          'ironfan-natty'
    bootstrap_distro    'ubuntu10.04-ironfan'
    chef_client_script  'client.rb'
    mount_ephemerals
  end

  environment           :dev

  role                  :systemwide
  cloud(:ec2).security_group :systemwide
  role                  :chef_client
  role                  :ssh
  cloud(:ec2).security_group(:ssh).authorize_port_range 22..22

  role                  :set_hostname

  role                  :volumes
  role                  :package_set,   :last

  role                  :org_base
  role                  :org_users
  role                  :org_final,     :last

  facet :raid do
    instances           1
    cloud(:ec2).flavor  'm1.large'
    recipe              'volumes::build_raid', :first

    cloud(:ec2) do
      availability_zones ['us-east-1c']
      vpc                 'vpc-fe8fca94'
      subnet              'subnet-a08ecbca'
      elastic_ip          '107.23.163.84'
      security_group      'raid-vpc' do
        authorize_port_range 22,"0.0.0.0/0",'tcp'
      end
    end

    cloud(:ec2).mount_ephemerals
    raid_group(:md0) do
      device            '/dev/md0'
      mount_point       '/raid0'
      level             0
      sub_volumes       [:ephemeral0, :ephemeral1]
    end
  end

  cluster_role.override_attributes({
    })
end
```

**Comment by gwilton** [comment]
It seems the issue lies in **lib/chef/knife/cluster_ssh.rb:54**. The c.machine.public_hostname is being used for cluster ssh. AWS forum explain that EIPs in VPC don't have a public_hostname set (https://forums.aws.amazon.com/message.jspa?messageID=406306). That means we should use the public_ipaddress or override it with the elastic_ip set in the cluster config.

``` ruby
        target = target.select {|t| not t.bogus? }
        addresses = target.map {|c| c.machine.public_hostname }.compact

        (ui.fatal("No nodes returned from search!"); exit 10) if addresses.nil? || addresses.length == 0

```

Not sure about this? Going to try a few different things on my end.

**Comment by UncleTallest** [comment]
Ahhh... Therein lies the problem and I should have noticed it right off. At present the code that does that actual attaching of  Elastic IPs does NOT handle the vpc domain or scope as written. 

This is by no means a solution, but I have the branching mechanism for some `vpc` versus `standard` domain issues in my personal fork; including VPC Elastic IPs as I recall. I've been working on some useful enhancements to my original Elastic IP code that wasn't anywhere near a finished state in time for the first release; nor was it in the original spec I was given. I haven't yet written any of the code to get the allocation_id from the requested address record itself yet, so I haven't initiated a pull request to Infochimps.

**Comment by UncleTallest** [comment]
About c.machine.public_hostname; I believe this is deeply related to issue https://github.com/infochimps-labs/ironfan/issues/195.

**Comment by gwilton** [comment]
Thanks @schade, going to checkout your Elastic_IP_enhanced branch and try it out. I can see there is a lot of work going into VPC/EIP, at the moment I might have to hack my way out of these issues. Will keep a watch on progress and provide any feedback on advancements made on my end.

**Comment by msaffitz** [comment]
I've hacked up a branch of ironfan that works reasonably well with VPCs at git: http://github.com/msaffitz/ironfan

After upgrading to 4.7.1, I'm also seeing the original issue reported here:

```
WARNING: Error running #<Ironfan::Broker::Computer(server=#<Ironfan::Dsl::Server(name="0", run_list_items=c{ role[systemwide], role[chef_client], role[ssh], role[volumes], role[package_set], role[minidash], role[tuning], role[zabbix_server], role[zabbix_cluster] }, clouds=c{ ec2 }, volumes=c{  }, environment=:production, cluster_role=#<Ironfan::Dsl::Role>, facet_role=#<Ironfan::Dsl::Role>, cluster_name="zabbix", facet_name="server")>, resources=c{ client, node, keypair, security_group__vpc-0e4fc567:default }, drives=c{ root, ephemeral0 }, providers=c{ chef, iaas })>:
WARNING: undefined method `group_id' for nil:NilClass
ERROR: undefined method `group_id' for nil:NilClass (NoMethodError)
/Users/msaffitz/.rbenv/versions/1.9.3-p194/lib/ruby/gems/1.9.1/bundler/gems/ironfan-3769b404eb11/lib/ironfan/provider/ec2/machine.rb:244:in `block in launch_description'
```

I wonder about the double underscore in `security_group__vpc-0e4fc567`

**Comment by UncleTallest** [comment]
I'm nearing something that's actually workable with no changes in usage; but was hit by this 'group_id' error as well... Still looking for the solution to that and then I'll have that pull request ready for the real world.

I looked askance at the double underscore there, but didn't delve very deeply when I first noticed it.

**Comment by temujin9** [comment]
The double underscore is intentional, to separate the resource type and resource identifier visually. Admittedly, it is a sign of code smell underneath, but not in itself a symptom of bugginess.

**Comment by UncleTallest** [comment]
Pull request https://github.com/infochimps-labs/ironfan/pull/233 submitted; let me know if testing finds anything needing more work. I will be continuing to add to this as time goes on.

**Comment by temujin9** [comment]
498f0c9 should fix the Elastic IP in VPC bug. The other (undefined method `group_id' ) seems to be an issue with earlier failure to actually create the VPC security group in question, and will require separate fix; please file a separate issue.
